File: 2-04-2024/Minimum Absolute Difference In BST.js
//{ Driver Code Starts

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split("\n").map(string => {
        return string.trim();
    });

    main();
});

function readLine() {
    return inputString[currentLine++];
}



class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

function buildTree(str) {
    // Corner Case
    if (str.length === 0 || str[0] === "N") return null;

    // Create the root of the tree
    let root = new Node(parseInt(str[0]));

    // Push the root to the queue
    let queue = [];
    let start = 0;
    queue.push(root);

    // Starting from the second element
    let i = 1;
    while (queue.length !== start && i < str.length) {

        // Get and remove the front of the queue
        let currNode = queue[start];
        start++;

        // Get the current node's value from the string
        let currVal = str[i];

        // If the left child is not null
        if (currVal !== "N") {

            // Create the left child for the current node
            currNode.left = new Node(parseInt(currVal));

            // Push it to the queue
            queue.push(currNode.left);
        }

        // For the right child
        i++;
        if (i >= str.length) break;
        currVal = str[i];

        // If the right child is not null
        if (currVal !== "N") {

            // Create the right child for the current node
            currNode.right = new Node(parseInt(currVal));

            // Push it to the queue
            queue.push(currNode.right);
        }
        i++;
    }

    return root;
}
//printing binary tree inorder form
function inorder(root) {
	let s = "";
	function inorderUtil(node){
	    if(!node){
	        return ;
	    }
	    inorderUtil(node.left);
	    s+=node.data;
	    s+=" ";
	    inorderUtil(node.right);
	}
	inorderUtil(root);
	console.log(s);
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++){
        
        let input_root = readLine().split(' ');
        let root = buildTree(input_root);
        
        let obj = new Solution();
        let res = obj.absolute_diff(root);
        console.log(res);
    }
}

// } Driver Code Ends




/*

class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}
*/

class Solution {
  //Your Code Start
    inorder(root, v) {
        if (!root) return;
        if (root.left) this.inorder(root.left, v);
        v.push(root.data);
        if (root.right) this.inorder(root.right, v);
    }

    absolute_diff(root) {
        const v = [];
        this.inorder(root, v);
        let ans = Infinity;
        const n = v.length;
        for (let i = 1; i < n; i++) {
            ans = Math.min(ans, v[i] - v[i - 1]);
        }
        return ans;
      //Your Code End
    }
}

        

File: 2-04-2024/Minimum Absolute Difference In BST.py
# class Node:
#     def __init__(self):
#         self.data = None
#         self.left = None
#         self.right = None
        
class Solution:
  #Your Code Start
    def inorder(self, root, v):
        if not root:
            return
        if root.left:
            self.inorder(root.left, v)
        v.append(root.data)
        if root.right:
            self.inorder(root.right, v)

    def absolute_diff(self, root):
        v = []
        self.inorder(root, v)
        ans = float('inf')
        n = len(v)
        for i in range(1, n):
            ans = min(ans, v[i] - v[i - 1])
        return ans
#Your Code End





#{ 
 # Driver Code Starts
from collections import deque


class Node:

    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


# Function to Build Tree
def buildTree(str):
    # Corner Case
    if len(str) == 0 or str[0] == 'N':
        return None

    # Creating list of strings from input string after splitting by space
    ip = str.split()

    # Create the root of the tree
    root = Node(int(ip[0]))

    # Push the root to the queue
    queue = deque()
    queue.append(root)

    # Starting from the second element
    i = 1
    while queue and i < len(ip):
        # Get and remove the front of the queue
        currNode = queue.popleft()

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if currVal != "N":
            # Create the left child for the current node
            currNode.left = Node(int(currVal))

            # Push it to the queue
            queue.append(currNode.left)

        # For the right child
        i += 1
        if i >= len(ip):
            break
        currVal = ip[i]

        # If the right child is not null
        if currVal != "N":
            # Create the right child for the current node
            currNode.right = Node(int(currVal))

            # Push it to the queue
            queue.append(currNode.right)

        i += 1

    return root


for _ in range(int(input())):
    s = input()
    root = buildTree(s)
    if root is None:
        continue
    if root.left is None and root.right is None:
        continue

    ob = Solution()
    print(ob.absolute_diff(root))

# } Driver Code Ends
